keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+-->
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+Currently, `kubelet --node-ip` has more-or-less no effect when using an
+external cloud provider. (It will require that the IP you pass is a
+known node IP, but it does not do anything else.) Assuming we change it
+to behave the same with external cloud providers as it does with legacy
+cloud providers and bare metal (ie, letting the user choose which node
+IP should be primary), then some users who were passing `--node-ip`
+before would need to stop passing it when upgrading, to avoid getting
+the new behavior. Perhaps this should warn and continue to use the old
+behavior during alpha, then warn and use the new behavior in beta.
+
+(Alternatively, since we need to extend `--node-ip` to support dual
+stack anyway, we could deprecate `--node-ip` and replace it with
+`--node-ips`, where the old `--node-ip` flag would continue to have the
+same semantics on external clouds until it went away.)
+
+Users on bare metal who upgrade and then convert their cluster from
+single-stack to dual-stack might need to switch back to single-stack
+first if they later needed to downgrade, due to the loss of some
+bare-metal-related dual-stack functionality when downgrading.
+
+External programs, scripts, etc that currently look at
+`Node.Status.Addresses` will eventually need to be updated to look at
+the new fields instead.
+
+### Version Skew Strategy
+
+`Node.Status.Addresses` can't actually go away until some hypothetical
+future `corev2.Node`, so older components will still be able to get
+all of the old information from there in a skewed cluster.
+
+During the skew period, components would not be able to assume
+unconditionally that `Node.Status.IPs` would be set, since they might
+be running against an old kubelet or cloud-controller-manager. (This
+includes kubelet itself, in the case of a new kubelet and an old
+cloud-controller-manager.)
+
+## Implementation History
+
+<!--
+Major milestones in the life cycle of a KEP should be tracked in this section.
+Major milestones might include
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+Instead of deprecating `Node.Status.Addresses`, we could just clarify
+its semantics instead. However, this would not fix the "API Cleanliness"
+problems.
+
+Additionally, if we want to make any changes to the semantics of the
+addresses (eg, "you shouldn't declare a single IP as both internal and
+external") then there would have to be some way for cloud providers to
+move from the old semantics to the new ones without breaking existing
```

Comment:
I'm not sure this is true.

Let's say we introduced new address types for `Status.Addresses`: `InternalIPv6` and `ExternalIPv6`. For single stack IPv6, the cloud provider could be smart enough to leave existing addresses that are using the old `InternalIP` and `ExternalIP`. We could actually control this behavior on the controller level so providers don't have to worry about it. 

For single-stack IPv4 and dual-stack, the new address types would be a natural transition.

Reply:
As an example, if a cloud provider returns a v6 address using the existing types:
```
- type: Hostname
  address: foobar
- type: InternalIP
  address: 2001:0db8:85a3:0000:0000:8a2e:0370
```

the cloud node controller (and kubelet?) can be updated to interpret the status as
```
- type: Hostname
  address: foobar
- type: InternalIP
  address: 2001:0db8:85a3:0000:0000:8a2e:0370
- type: InternalIPv6
  address: 2001:0db8:85a3:0000:0000:8a2e:0370
```